torch/csrc/dynamo/guards.cpp

```diff
@@ -142,6 +149,7 @@ class TensorCheck {
   // necessarily capture device indices correctly.
   at::DeviceIndex device_index_;
   bool requires_grad_;
+  bool grad_mode_enabled_;
```

Comment:
Do we actually need grad_mode_enabled_ here? It seems wholly redundant w GRAD_MODE (I know this is a prior issue with the code, and not something you introduced).

Would you be open to pulling it out?  the `GRAD_MODE` should cover all cases here.

Reply:
yep happy to pull it out.

I see there's already a `GRAD_MODE` guard defined here. Does that mean it's enough just to yank this bool and only guard on `requires_grad` as part of the tensor guard? Or do you suspect there's something else we need. I can also yank it and see if my new tests still pass.

Reply:
>  Does that mean it's enough just to yank this bool and only guard on requires_grad as part of the tensor guard? 

That should suffice. If not, we need to figure out why `GRAD_MODE` isn't working.

Reply:
The tests fail when I remove it.

> If not, we need to figure out why GRAD_MODE isn't working.

It looks like the only use of the `GRAD_MODE` guard is in `GradModeVariable` [here](https://github.com/pytorch/pytorch/blob/96f46316c9d4788adbfdff531d0c35ecdcf316e9/torch/_dynamo/variables/ctx_manager.py#L114). That will probably be created if we see a `no_grad()` block **inside** the compiled region. But it looks like we aren't guarding on whatever the current grad state was when we first started compiling. Does that seem like the right thing to go fix (and any pointers?)

Reply:
You can just unconditionally add a GRAD_MODE guard at the start of tracing in output_graph.py, similar to GuardBuilder.DETERMINISTIC_ALGORITHMS

Reply:
thanks for the pointer! Updated, tests now pass